### PR TITLE
Remove Material.h from ObjComponent.h and Object.h

### PR DESCRIPTION
**Pull request by peterfpeterson**
It was discovered that `Material.h` was included directly into `Object.h` and `ObjComponent.h`. Half of the fix was to have `ObjComponent` use its `Object->sample` to hold the material rather than explicitly have its own copy, The other half was to return `Material` by value rather than reference since they are relatively small. The effect of this work is to reduce compile times and source file dependencies.

**To test:**

_There was no associated issue_

_Since this is not a user-facing change, it does not need to be in the release notes._

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?
##### Functional Tests
- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

This is part of the ongoing effort to decrease build times.

### CODE REVIEW

**Comment by martyngigg** [comment]
Do we need this in `3.7.2`?

**Comment by peterfpeterson** [comment]
I must have just selected the nearest milestone. Fixed.

File: Framework/Geometry/inc/MantidGeometry/Instrument/SampleEnvironmentSpecParser.h [hunk 1]
```diff
@@ -3,6 +3,7 @@
 
 #include "MantidGeometry/DllConfig.h"
 #include "MantidGeometry/Instrument/SampleEnvironmentSpec.h"
+#include "MantidKernel/Material.h"
```
Comment (martyngigg) [inline Framework/Geometry/inc/MantidGeometry/Instrument/SampleEnvironmentSpecParser.h]: Was this necessary? I notice it was also included in the `.cpp` too?

**Comment by peterfpeterson** [comment]
MSVS doesn't seem to like a `std::unordered_map<string, unique_ptr<Material>)`. Do you know how to get around it?

**Comment by martyngigg** [comment]
Maybe it's not possible with `unique_ptr` and the `std` implementation of `unordered_map`. I can't actually get it to work with `gcc` at https://gcc.godbolt.org/. As the `SampleEnvironmentParser` is not low level I think we'll have to live with the header file there.

**Comment by martyngigg** [comment]
Everything looks good here now.

:shipit:

**Comment by FedeMPouzols** [comment]
Nice. This reduced the average number of includes in Framework cpp files from 116.6 down to 114.9.

**Pull request merged**